Node selector

Ranelim opened the pull request:
Introducing Node-selector as it appears in-page and the Edit modal.

@openshift/team-ux-leads

lizsurette commented on designs/virtualization/4.5/Scheduling/Node-selector/node-selector.md:
```diff
@@ -0,0 +1,37 @@
+![Node selector modal ](img/Node-selector-1-3.jpg)
+In case no match found, the Nodes checker will fire a warning feedback.
+In this case, the user has added another label and no nodes found that have both the first and the second label.
+At this point the user might click the "explore nodes list" external link to search for a node of a more suited label.
```
```suggestion
At this point the user might click the "explore nodes list" external link (which will open the full Nodes list in a new tab) to search for a node of a more suited label.
```

andybraren commented on designs/virtualization/4.5/Scheduling/Node-selector/node-selector.md:
```diff
@@ -0,0 +1,37 @@
+![Node selector modal ](img/Node-selector-1-1.jpg)
+On typing some value within any of the fields, the Node Checker will check for nodes with matching labels, with a time out of [1] second to kick in.
+
+![Node selector modal ](img/Node-selector-1-2.jpg)
```
Nit or maybe an FYI: We're probably not super consistent with this elsewhere in the Console (maybe we should all agree on a convention) but at least in Overviews/Dashboards we standardized on using the word "View" for things that will take the user somewhere else (like "View details"). Related to that, maybe "Explore nodes list" should be "View nodes list".

"See" might be a better word to reveal the popover. "Show" is also possible, but at least OperatorHub uses that word to indicate that more things will show up inline when clicked.

andybraren commented on designs/virtualization/4.5/Scheduling/Node-selector/node-selector.md:
```diff
@@ -0,0 +1,37 @@
+
+![Node selector modal ](img/Node-selector-1-2.jpg)
+In case a match found, The Node Checker will fire a positive feedback.
+It will also include a link to view the matching nodes within a popover, on hover or on click.
```
```suggestion
It will also include a link to view the matching nodes within a popover on click.
```

I believe popovers like this are always on click, at least out of the box in PatternFly. Dark-colored [tooltips](https://www.patternfly.org/v4/documentation/react/components/tooltip) appear on hover, but we should try to use those rarely.

andybraren commented on designs/virtualization/4.5/Scheduling/Node-selector/node-selector.md:
```diff
@@ -0,0 +1,37 @@
+Clicking on the "O labels" link text will open that modal as well.
+
+
+![Node selector modal ](img/Node-selector-1-0.jpg)
```
I'm still unsure about using inline PatternFly alert styling for these waiting & checking states that aren't actually alerts. The grey color here would also need to be custom CSS, and the in-progress icon for this "waiting" state doesn't seem typical.

We definitely should signal to the user that this kind of validation will happen, but hopefully we can do so without using an inline alert. Maybe something like a dedicated "Matching nodes" section that just uses plain text until we reach the success/warning inline alert states:

![matching-nodes](https://user-images.githubusercontent.com/9122899/77130602-f59d7b80-6a2e-11ea-9200-db819b3acba1.png)

Ranelim replied:
I made, the changes u suggested. I think this not as good as having the entire Nodes checker in a box. Giving the user some visual anchor and helping her understand that some feedback is here. 
But I think we can move forward with this implementation fo now for the sake of the rest of the PRs depending on this one.

WDYT?

matthewcarleton replied:
I'm not 100% sure we can't use an inline info alert here. It feels like something we want to notify the user of and bring their attention to. I admit though this is somewhat uncharted territory in terms of inline alerts.
@Ranelim correct me if I'm wrong but this text is somewhat inaccurate
 "Please add at least one key/value pair above to check for matching nodes" 
I don't believe you have to add both key/value, you could just add just key without the value right?
If so we should tweak this text a bit to cover that.

Ranelim replied:
@matthewcarleton 


> @Ranelim correct me if I'm wrong but this text is somewhat inaccurate
> "Please add at least one key/value pair above to check for matching nodes"
> I don't believe you have to add both key/value, you could just add just key without the value right?

You are right. It went under my radar. This topic came up in our last Stakeholder meeting:
This is not actually a label. It could be a full label and it could be just a (full) Key and "Exist/nonExist" value. So we need to make modifications in the fields to support and surface that, as well as aligning this text.

For now, I want to share an alternative design to this Nodes Checker that uses "Info" notification instead of any custom design:

![Node-Checker-03](https://user-images.githubusercontent.com/48942365/77430014-2ee53b00-6de3-11ea-9d0a-dca0dc39d28d.jpg)

@matthewcarleton @andybraren 
WDYT?

matthewcarleton replied:
I think using an info alert here rather than a generic bell icon might be more appropriate. It might also be nice to add the spinner in the content area of the alert. WDYT? If you stick with the bell here I'd suggest using the pf colour rather than the info colour for the generic alert.

andybraren replied:
> I think this not as good as having the entire Nodes checker in a box. Giving the user some visual anchor and helping her understand that some feedback is here.

In general I think the [usage and behavior guidelines](https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/alerts-and-notifications) should determine which component is used and ultimately how things look, rather than the other direction. I totally see the appeal of making this area look like one "widget" in a box though. See below.

> I'm not 100% sure we can't use an inline info alert here. It feels like something we want to notify the user of and bring their attention to. I admit though this is somewhat uncharted territory in terms of inline alerts.

Maybe my concern comes down to the wording/framing of the waiting and checking states.

As-is, I'm not sure "Waiting for input to check matching nodes" aligns with the [guidelines](https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/alerts-and-notifications)' suggestion to show alerts "as the result of a user action within a page." The "checking" state at least depends on user input/action. The wording in those guidelines admittedly has some wiggle room though. For some reason they only recommend placing inline alerts at the top of the page/content, but that's hardly followed. 🤷‍♂️

Maybe making the text a bit more "informational" and explanatory would help here. WIP, but something like:

Waiting state (info alert styling with the info icon):
> **Matching nodes will be searched for**
> Please add at least one label key above to search for nodes that match this selector.

Checking state (possibly the default alert styling with the bell, not sure):
> **Searching for matching nodes...**
> Please wait while nodes that match this selector are found.

matthewcarleton replied:
ya I agree we are pushing on alerts here a bit, but I agree it's not 100% clear and we can probably wiggle a bit here :) 
This could be a case for a unique component  b/c it has multiple states. I'm sure we could see this used in other places and having a recognizable pattern for dynamic results could be very nice. 
I think those suggestions work. I'm curious too how long we are talking for results to show. It could be almost instant, in that case we maybe don't see the searching state at all.

Ranelim replied:
> As-is, I'm not sure "Waiting for input to check matching nodes" aligns with the guidelines' suggestion to show alerts "as the result of a user action within a page."

I believe that infobox resolves that issue.

Adding two more options:
I think that if we are having a hard time with the Info notification box switching to alert, then the second option of support text is good as well. In fact, I'm starting to think this is better because it is not a notification. It pushes the right "volume" of attention and is not in par with the actual results.
As andy suggested, the intention was to create this "Smart widget box". But since it is perceived as a Notification and we cannot create something unique there is no point in mimicking an info notification pattern.

@andybraren @matthewcarleton  
WDYT

![Node-Checker-01](https://user-images.githubusercontent.com/48942365/77513159-2a6d6080-6e7d-11ea-82b3-0e8902e78c89.jpg)
![Node-Checker-02](https://user-images.githubusercontent.com/48942365/77513172-2e00e780-6e7d-11ea-8bcd-1db2b5fb87b8.jpg)

Ranelim replied:
@andybraren @matthewcarleton 
WDYT?
![Node-Checker-03](https://user-images.githubusercontent.com/48942365/77552816-af2b9f00-6ebc-11ea-9b14-91462fd3a26e.jpg)

matthewcarleton replied:
I like this! Seems like a nice combination of all concerns.

andybraren replied:
Agreed, thank you for tolerating me. 😄

The descriptive text might also be a good opportunity to explain what the Node Selector is/does. Are "node labels or node keys" referring to the same thing here?

Ranelim replied:
> The descriptive text might also be a good opportunity to explain what the Node Selector is/does. Are "node labels or node keys" referring to the same thing here?

Yes, the user may enter entire label or just the Key of that label (to tap all the nodes with that Key)

andybraren commented on designs/virtualization/4.5/Scheduling/Node-selector/node-selector.md:
```diff
@@ -0,0 +1,37 @@
+Introducing Node-selector as it appears in-page and the Edit modal.
+
+![Node selector field in the Scheduling section ](img/Node-selector-0-0.jpg)
+The node-selector field is pulled from the Workload details section and placed under the Scheduling section.
```
Nit: "Node Selector" and all other field headings should be capitalized.

spadgett commented:
We need to decide how to handle the case where the user doesn't have permissions to list nodes. Most users won't.

Is there a reason we can't use our existing components for showing label selectors and editing labels? I'd think we want to use those for consistency with other pages and dialogs. We had an existing node selector dialog that would be simple to bring back (vs writing an entirely new components from scratch).

spadgett commented:
This is how we show label selectors elsewhere in the UI. It links to the Search page with a filtered list, which lets you see what would be matched. Node selectors seem really important to show when set and they usually are short enough just to inline.

<img width="991" alt="Screen Shot 2020-03-20 at 9 26 39 AM" src="https://user-images.githubusercontent.com/1167259/77167767-0a106100-6a8d-11ea-8ad6-c17d2cced3a9.png">

Ranelim commented:
> We need to decide how to handle the case where the user doesn't have permission to list nodes. Most users won't.

Currently, we think we shouldn't display any nodes/nod-label content to users with no permissions 
This is what we learned (with the help of @jelkosz and @fdeutsch@redhat.com:

- Users with no permission to view nodes, do not have permission to view the node labels as well. Those labels sometimes contain sensitive info that should be hidden. 
- For that reason, those users can't view that content anyway or modify it. 
- Those users can't really articulate label based requirements since they don't know what those labels or nodes are, so there is no point to display any node-label based requirements to them
- We also learned that for users with no node-permission, such as the Dev console, the console itself does not have node-labels in the back-end (sorry for the awkward terminology). So it is not possible, to begin with.

- 

> 
> Is there a reason we can't use our existing components for showing label selectors and editing labels? I'd think we want to use those for consistency with other pages and dialogs. We had an existing node selector dialog that would be simple to bring back (vs writing entirely new component from scratch).

Today, we have more than one UI to view label-like components in-page and more than one UI to edit those in modal.
Labels are presented in-page and edited in one free text box in a modal
Node labels are presented in-page as well (in purple) but presented differently in the current node selector. 
annotations, which are not labels but are constructed the same way, are presented in a link. That link opens up a modal that presents a Key/Value field pair. I don't think there is any need for a different experience for annotations, as a better visual indication.

This design is meant to align these into one behavior. The following models that will appear in the following PRs will reuse the same components and the same UI. So, it is an attempt to create some consistency in the Labels/Node-labels/Node-selectors/Annotations.
It does not include Annotations but suggests we will modify annotations as well.

spadgett commented:
> Currently, we think we shouldn't display any nodes/nod-label content to users with no permissions

That's fine. I guess my point is that few users will be able to take advantage of these changes :/

> This design is meant to align these into one behavior. The following models that will appear in the following PRs will reuse the same components and the same UI. So, it is an attempt to create some consistency in the Labels/Node-labels/Node-selectors/Annotations.
> It does not include Annotations but suggests we will modify annotations as well.

We're actually extremely consistent throughout the UI in how we display each. Labels are always displayed one way for all resources. Label selectors are always displayed one way. Annotations are always displayed one way.

Each is fundamentally different and has different design considerations.

* Labels are meant to be set by users. They tend to be short values. They link to the Search page. The color matches the resource badge color.
* Label selectors aren't always name-value pairs. They can be more complicated expressions. They match other resources and link to the Search page. The color is the color of the *target* resource. The magnifying glass indicates that you're *searching* for these labels on another resource.
* Annotations tend to be internal to system and can be very long values like JSON blobs. It's not as common for the user to edit these, although occasionally they need to. Usually you don't need to see these values as a user. Annotations can't be searched on, so there's no Search page link.

Treating these exactly the same would be a significant step back in usability in my opinion.

Ranelim commented:
@spadgett 
I've updated the node selector in-page placement to match what we have today.
I've also added an "Operator" drop-down between the Key and Value fields, within the modal.
The Operator drop-down options are:
- (key) 'Equals' (value)
- (key) 'Exist' (value is disabled)
- (key) 'notExist' (value is disabled)

I really couldn't find an example of the 'NotExist' so it is a guess.
The examples I did find for the 'Exist' case is that within the YAML, the Value is simply not mentioned. 
For example: **some-random-key: ' '**

Because I couldn't find an example of 'NotExist', the In-page presentation of it is a guess as well (**some-random-key=notExist**)

Should I stick with it or just drop the 'NotExist case'. If to drop it, we can drop the Operator drop-down and simply apply the 'Exist' case by not typing any Value, like in the YAML.

WDYT?

glekner commented:
<img width="448" alt="Screen Shot 2020-03-30 at 13 59 39" src="https://user-images.githubusercontent.com/24938324/77905329-bc66d600-728e-11ea-858e-6323de306686.png">

I notice the Selector Component doesn't wrap the label with a filled box, should we change it @spadgett

spadgett commented:
> @spadgett
> I've updated the node selector in-page placement to match what we have today.
> I've also added an "Operator" drop-down between the Key and Value fields, within the modal.

@Ranelim Node selectors specifically are only key-value pairs. They don't have an operator. (Other selectors can like pod selectors on a replica set, but node selectors don't.)

spadgett commented:
OK, looks like you dropped this from the PR already, thanks!
